src/gather.cu

```diff
@@ -31,10 +31,12 @@ testResult_t GatherInitData(struct threadArgs* args, ncclDataType_t type, ncclRe
   size_t sendcount = args->sendBytes / wordSize(type);
   size_t recvcount = args->expectedBytes / wordSize(type);
   int nranks = args->nProcs*args->nThreads*args->nGpus;
+  int nGpusVisible;
+  CUDACHECK(cudaGetDeviceCount(&nGpusVisible));
 
   for (int i=0; i<args->nGpus; i++) {
     int gpuid = args->localRank*args->nThreads*args->nGpus + args->thread*args->nGpus + i;
-    CUDACHECK(cudaSetDevice(gpuid));
+    CUDACHECK(cudaSetDevice(gpuid % nGpusVisible));
     int rank = ((args->proc*args->nThreads + args->thread)*args->nGpus + i);
```

Comment:
Why do we need those modifications to each test?

Reply:
These are required when `CUDA_VISIBLE_DEVICES` are set, right? From the [documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g18808e54893cfcaafefeab31a73cc55f) - 1) for `cudaSetDevice`,

````Sets device as the current device for the calling host thread. Valid device id's are 0 to ([cudaGetDeviceCount()]```

 Since, `CUDA_VISIBLE_DEVICES=X` would lead to `cudaDeviceCount` set to 1, we will require this change in every collective to set the right GPU for kernel operations.

Reply:
Understood. That code is old; we should not re-compute the gpuid all around the code and store it in `args`. Also it's a bit of a hack to use modulo. It works for your use cases but it has other side effects.
For example, if you run with 2 ranks, and set CUDA_VISIBLE_DEVICES to "0,1,2,3" on one rank and "4,5,6,7" and the other rank, then launch with only 2 GPUs per rank, users would expect to use 0,1 and 4,5 but instead we'd use 0,1 and 6,7.
Also, today if you run on a 4 GPUs system and launch 8 ranks or one rank with `-g 8` it will error out as "invalid CUDA numeral" or something similar. With this patch it would try to re-use each GPU twice (and generate and error later during NCCL init).

I think the reason for using another environment variable (e.g. `NCCL_TESTS_DEVICE`) is to indicate whether the local ranks should work within the global `CUDA_VISIBLE_DEVICES` and ensure they use different GPUs within that set (default) or consider the visible GPUs are dedicated to them and start from GPU 0. We could compute the GPUs are `NCCL_TESTS_DEVICE`+i and that would probably work as well.